ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
-    /// with zero operands until they are explicitly filled in.
-    NotPresent = 0,
-    /// The operand references a previously defined local variable.
-    Local,
+/// A structure type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct StructType {
+    // FIXME: todo.
 }
 
-impl From<u64> for OpKind {
-    fn from(v: u64) -> Self {
-        // unwrap safe assuming only valid discriminant numbers are used.
-        Self::from_repr(v).unwrap()
-    }
+/// A function type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct FuncType {
+    /// Type indices for the function's formal arguments.
+    arg_tys: Vec<TypeIndex>,
+    /// Type index of the function's return type.
+    ret_ty: TypeIndex,
+    /// `true` if the function has varargs.
+    has_varargs: bool,
 }
 
-#[derive(Debug, FromRepr, PartialEq)]
-#[repr(u64)]
-pub enum OpCode {
-    Load,
-    LoadArg,
+impl FuncType {
+    pub(crate) fn new(arg_tys: Vec<TypeIndex>, ret_ty: TypeIndex, has_varargs: bool) -> Self {
+        Self {
+            arg_tys,
+            ret_ty,
+            has_varargs,
+        }
+    }
+
+    pub(crate) fn num_args(&self) -> usize {
+        self.arg_tys.len()
+    }
 }
 
-impl From<u64> for OpCode {
-    fn from(v: u64) -> Self {
-        // unwrap safe assuming only valid discriminant numbers are used.
-        Self::from_repr(v).unwrap()
+// FIXME: macro generate `new` and `get` for all index types.
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+pub(crate) struct InstrIndex(u16);
+
+impl InstrIndex {
+    pub(crate) fn new(v: usize) -> Self {
+        // FIXME: propagate error
+        Self(u16::try_from(v).unwrap())
+    }
+
+    pub(crate) fn get(&self) -> u16 {
+        self.0
     }
 }
 
-#[derive(Debug, PartialEq)]
-pub enum Operand {
-    Long(LongOperand),
-    Short(ShortOperand),
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) struct FuncIndex(u16);
+
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub(crate) struct ExtraArgsIndex(u16);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub(crate) struct TypeIndex(u16);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) struct ConstIndex(u16);
+
+impl ConstIndex {
+    pub(crate) fn get(&self) -> u16 {
+        self.0
+    }
 }
 
-impl Operand {
-    pub(crate) fn new(kind: OpKind, val: u64) -> Self {
-        // check if the operand's value can fit in a short operand.
-        if val <= SHORT_OPERAND_MAX_VALUE {
-            Self::Short(ShortOperand::new(kind, val))
-        } else {
-            todo!()
+/// The packed representation of an instruction operand.
+///
+/// # Encoding
+///
+/// ```ignore
+///  1             15
+/// +---+--------------------------+
+/// | k |         index            |
+/// +---+--------------------------+
+/// ```
+///
+///  - `k=0`: `index` is a local variable index
+///  - `k=1`: `index` is a constant index
+///
+///  The IR can represent 2^{15} = 32768 locals, and as many constants.
```

Comment:
I think it can represent `2^15-1` i.e. 32767.

Reply:
We are counting how many locals and constants we can represent. Index 32767 is the 32768th index, as indices start from 0. Or am I being silly?

Reply:
Ah, sorry, I meant `2^15-3` because of the "reserved" types. Not sure if that's worth mentioning here though.

Reply:
The reserved types are still types that should be accounted for thought, right? I don't think we need to make any distinction between reserved and other types.